### PR TITLE
Never send an empty content block

### LINKED ISSUE #762: `litellm.completion` errors with prompt cache metadata for empty assistant message
This came up in a long-running session this afternoon. Will figure out and add repro details later.

Note: this may be specific to the first-party Anthropic API.

### PR DESCRIPTION
**Pull request by eb8680**
Resolves #762 

This PR prevents the `handlers.llm` harness from constructing or accepting messages with empty content blocks, which provider APIs (especially Anthropic) don't like and don't respond to uniformly. This also prevents the failure in #762 from happening.

### CODE REVIEW

File: effectful/handlers/llm/harness/durability/transaction.py [hunk 1]
```diff
@@ -36,12 +39,18 @@ def get_history(cls) -> collections.abc.MutableSequence[Message]:
     def append_message(cls, message: Message) -> None:
         """Append `message` to the ambient history, if it is legal where it lands.
 
-        Both checks are about position rather than content, which is why they sit
-        here: every message the harness records passes through this method,
-        including the ones a failed attempt records on its way out, and those are
-        the ones that get a history into a shape no provider will accept.
+        Two of the checks are about position rather than content, which is why
```
Comment (jfeser) [inline effectful/handlers/llm/harness/durability/transaction.py]: The claude comments are a rough read. I don't know if there's an easy way to get claude to use normal english, but I've had better luck with getting chatgpt to produce reasonable text.

File: effectful/handlers/llm/harness/hooks.py [hunk 2]
```diff
@@ -267,7 +267,10 @@ def call_assistant[T](
     result = None
     if not tool_calls:
         serialized_result = message.get("content") or message.get("reasoning_content")
-        assert isinstance(serialized_result, str)
+        assert isinstance(serialized_result, str) and serialized_result.strip(), (
```
Comment (jfeser) [inline effectful/handlers/llm/harness/hooks.py]: I don't know if we can assert that model results are nonempty. Empty results should be retried like other invalid outputs.

File: effectful/handlers/llm/harness/serialization.py
Comment (jfeser) [inline effectful/handlers/llm/harness/serialization.py:179]: Per chatgpt, if we change the semantics of `to_content_blocks` to return an empty list when `encoded_obj == ''`, we lose output in some cases where formatting would have produced non-empty content (e.g. `{value:r}`).

File: effectful/handlers/llm/harness/serialization.py [hunk 3]
```diff
@@ -113,6 +116,18 @@ def walk(v: typing.Any) -> None:
     return blocks
 
 
+def _is_empty_text_block(block: typing.Any) -> bool:
```
Comment (jfeser) [inline effectful/handlers/llm/harness/serialization.py]: This test is more restrictive than the test for empty blocks in `to_content_blocks`, so `to_content_blocks` will still return blocks that fail this.

**Review by jfeser — CHANGES_REQUESTED** [review]
A few chatgpt comments:

File: effectful/handlers/llm/harness/hooks.py (hunk 2 above)
Comment (eb8680) [reply effectful/handlers/llm/harness/hooks.py]: They are retried within `completion` when `completion` is given `num_retries > 0` as a kwarg, which the harness does by default via `LiteLLMConfigurer`. If an empty result survives to this point it's almost certainly a persistent error with the API that doesn't have a client-side workaround and should thus fail loudly.

Comment (jfeser) [reply effectful/handlers/llm/harness/hooks.py]: `TenacityRetryer` only catches `ToolDecodingError` and `ResultDecodingError`. I see that we convert `Exception` to these in a few places, but I don't think this assert is inside one of these converting `try-except`s.

File: effectful/handlers/llm/harness/serialization.py
Comment (eb8680) [reply effectful/handlers/llm/harness/serialization.py:179]: OK that should be fixed

File: effectful/handlers/llm/harness/durability/transaction.py (hunk 1 above)
Comment (eb8680) [reply effectful/handlers/llm/harness/durability/transaction.py]: I've de-slopped, I can't get it to stop doing that no matter how much I berate it...

File: effectful/handlers/llm/harness/hooks.py (hunk 2 above)
Comment (eb8680) [reply effectful/handlers/llm/harness/hooks.py]: I'm not sure about putting this assertion under the retry loop because I think the only way it can arise apart from some opaque server-side error is from context truncation, but I've put it there anyway and we can see if it causes any problems down the line.

File: effectful/handlers/llm/harness/serialization.py (hunk 3 above)
Comment (eb8680) [reply effectful/handlers/llm/harness/serialization.py]: Fixed, the same predicate is used everywhere this check is applied

File: effectful/handlers/llm/harness/durability/transaction.py (hunk 1 above)
Comment (jfeser) [reply effectful/handlers/llm/harness/durability/transaction.py]: Yeah, I think it's a fundamental issue with Claude.
